nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -2549,184 +2707,184 @@ impl DataStore {
+            // now.
+            return Ok(VolumeReplaceResult::ExistingVolumeDeleted);
         }
-        let err = OptionalError::new();
 
-        let conn = self.pool_connection_unauthorized().await?;
-        self.transaction_retry_wrapper("volume_replace_region")
-            .transaction(&conn, |conn| {
-                let err = err.clone();
-                async move {
-                    // Grab the old volume first
-                    let maybe_old_volume = {
-                        volume_dsl::volume
-                            .filter(volume_dsl::id.eq(existing.volume_id))
-                            .select(Volume::as_select())
-                            .first_async::<Volume>(&conn)
-                            .await
-                            .optional()
-                            .map_err(|e| {
-                                err.bail_retryable_or_else(e, |e| {
-                                    VolumeReplaceRegionError::Public(
-                                        public_error_from_diesel(
-                                            e,
-                                            ErrorHandler::Server,
-                                        )
-                                    )
-                                })
-                            })?
-                    };
+        let old_vcr: VolumeConstructionRequest =
+            match serde_json::from_str(&old_volume.data()) {
+                Ok(vcr) => vcr,
+                Err(e) => {
+                    return Err(err.bail(ReplaceRegionError::SerdeError(e)));
+                }
+            };
 
-                    let old_volume = if let Some(old_volume) = maybe_old_volume {
-                        old_volume
-                    } else {
-                        // Existing volume was hard-deleted, so return here. We
-                        // can't perform the region replacement now, and this
-                        // will short-circuit the rest of the process.
+        // Does it look like this replacement already happened?
+        let old_region_in_vcr =
+            match region_in_vcr(&old_vcr, &existing.region_addr) {
+                Ok(v) => v,
+                Err(e) => {
+                    return Err(
+                        err.bail(ReplaceRegionError::RegionReplacementError(e))
+                    );
+                }
+            };
+        let new_region_in_vcr =
+            match region_in_vcr(&old_vcr, &replacement.region_addr) {
+                Ok(v) => v,
+                Err(e) => {
+                    return Err(
+                        err.bail(ReplaceRegionError::RegionReplacementError(e))
+                    );
+                }
+            };
 
-                        return Ok(VolumeReplaceResult::ExistingVolumeDeleted);
-                    };
+        if !old_region_in_vcr && new_region_in_vcr {
+            // It does seem like the replacement happened
```

Comment:
You mean

```rust
        if !old_region_in_vcr && new_region_in_vcr {
            // It does seem like the replacement happened
            return Ok(VolumeReplaceResult::AlreadyHappened);
        } else if old_region_in_vcr && !new_region_in_vcr {
            // Replacement hasn't happened yet
        } else if old_region_in_vcr && new_region_in_vcr {
            // XXX wat
        } else if !old_region_in_vcr && !new_region_in_vcr {
            // XXX wat
        }
```
?

Reply:
Yes.

It seems like `old_region_in_vcr && !new_region_in_vcr ` is what we expect, and the other two are things that should not happen?

Reply:
Right, yeah. `old_region_in_vcr && new_region_in_vcr` means we performed a replacement (because new is there) but old is still there, meaning that the replacement couldn't have taken place, and `!old_region_in_vcr && !new_region_in_vcr` means that the thing we're being asked to replace isn't there to begin with, and a replacement hasn't taken place (because new isn't there).

Do you think I should have those cases in the code?

Reply:
We chatted and I did put some comments in the code in f52fa636f